pkg/auth/tokens/sha256.go

```diff
@@ -0,0 +1,68 @@
+package tokens
+
+import (
+	"bytes"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+const hashFormat = "$%d:%s:%s:%s" // $version:algorithm:salt:hash -> $1:SHA256:abc:def
+const Version = 1
+const Algorithm = "SHA256"
+
+// CreateSHA256Hash can be used for basic key hashing, includes a random salt
+func CreateSHA256Hash(secretKey string) (string, error) {
+	salt := make([]byte, 8)
+	_, err := rand.Read(salt)
+	if err != nil {
+		return "", err
+	}
+	hash := sha256.Sum256([]byte(fmt.Sprintf("%s%s", salt, secretKey)))
+	encSalt := base64.RawStdEncoding.EncodeToString(salt)
+	encKey := base64.RawStdEncoding.EncodeToString(hash[:])
+	return fmt.Sprintf(hashFormat, Version, Algorithm, encSalt, encKey), nil
+}
+
+// VerifySHA256Hash takes a key and compares it with stored hash, including its salt
+func VerifySHA256Hash(hash, secretKey string) error {
+	if !strings.HasPrefix(hash, "$") {
+		return errors.New("hash format invalid")
+	}
+	splitHash := strings.SplitN(strings.TrimPrefix(hash, "$"), ":", 4)
+	if len(splitHash) != 4 {
+		return errors.New("hash format invalid")
+	}
+
+	version, err := strconv.Atoi(splitHash[0])
+	if err != nil {
+		return err
+	}
+	if version != Version {
+		return fmt.Errorf("hash version %d does not match package version %d", version, Version)
+	}
+
+	alg, salt, enc := splitHash[1], splitHash[2], splitHash[3]
+	if alg != Algorithm {
+		return fmt.Errorf("hash algorithm %s does not match package algorithm %s", alg, Algorithm)
+	}
+	// base64 decode stored salt and key
+	decodedKey, err := base64.RawStdEncoding.DecodeString(enc)
+	if err != nil {
+		return err
+	}
+	decodedSalt, err := base64.RawStdEncoding.DecodeString(salt)
+	if err != nil {
+		return err
+	}
+	// compare the two
+	hashedSecretKey := sha256.Sum256([]byte(string(decodedSalt) + secretKey))
+	if !bytes.Equal(decodedKey, hashedSecretKey[:]) {
```

Comment:
This "should never happen" but Darren requested a sanity check here that both values are non-zero length, so that there's no way a mistake leads to auth being `if ( "" == "")` and always succeeding

Reply:
https://golang.org/pkg/crypto/subtle/#ConstantTimeCompare

Reply:
updated both